### ISSUE TITLE
Warning: count(): Parameter must be an array or an object that implements Countable 

### ISSUE THREAD

**Issue by wpsmort** [semperfiwebdesign/all-in-one-seo-pack#1491]
Reported in both free and Pro forums here:

https://semperplugins.com/support/troubleshooting-subscribe2-html/error-on-php-7-2-parameter-must-be-an-array-or-an-object/
https://wordpress.org/support/topic/getting-warning-after-installing-all-in-one-seo-pack/

Warning: count(): Parameter must be an array or an object that implements Countable in /wp-content/plugins/all-in-one-seo-pack-pro/aioseop_class.php on line 2643

Happens on a page or post. See screenshot:

![aioseo-php72-error_preview](https://user-images.githubusercontent.com/17548525/35066102-7ec0e5f8-fb9d-11e7-8958-bd2b3f15687c.png)

**Comment by EkoJr** [comment]
I have the same error to I believe. Here's the Call Stack as well.

```
Call Stack
--

1  | 0.0007 | 439400   | {main}( )                                      | ...\post.php:0
2  | 0.3448 | 24443752 | include( 'C:\xampp\htdocs\aioseop\wp-admin\edit-form-advanced.php' ) | ...\post.php:174
3  | 0.5023 | 25216296 | do_meta_boxes( )                               | ...\edit-form-advanced.php:714
4  | 0.5026 | 25216296 | All_in_One_SEO_Pack->display_tabbed_metabox( ) | ...\template.php:1063
5  | 0.5027 | 25216888 | All_in_One_SEO_Pack->display_metabox( )        | ...\aioseop_class.php:4470
6  | 0.5027 | 25216888 | All_in_One_SEO_Pack->display_options( )        | ...\aioseop_module_class.php:2491
7  | 0.5035 | 25228432 | apply_filters( )                               | ...\aioseop_module_class.php:2347
8  | 0.5035 | 25228728 | WP_Hook->apply_filters( )                      | ...\plugin.php:203
9  | 0.5035 | 25229720 | All_in_One_SEO_Pack->filter_settings( )        | ...\class-wp-hook.php:286
10 | 0.5036 | 25229768 | All_in_One_SEO_Pack->get_page_snippet_info( )  | ...\aioseop_class.php:3348
11 | 0.5178 | 25241008 | All_in_One_SEO_Pack->aiosp_mrt_get_url( )      | ...\aioseop_class.php:1403
```

**Comment by EkoJr** [comment]
@wpsmort @michaeltorbert This is an easy fix, but there's 2 different ways to go about this, and I am unsure if there may be a reason for checking the query variable directly.

**Method 1**
Uses the same concept as before, but checks the value.
```
$haspost = false;
if ( ! empty( $query->posts ) ) {
	$haspost = count( $query->posts ) > 0;
}
```
**Method 2**
Uses the built-in function, and would be cleaner.
```
// OR should it be using a WP function; have_posts()?
$hasposts = have_posts();

**Comment by michaeltorbert** [comment]
If there's a WP function, I don't know why we wouldn't use that. Maybe @contactashish13 can weigh in here?

BTW, whenever introducing a new WP function to AIOSEOP, always check the version it was created. Currently, we support back to WP 4.0. If you need a newer function, you'll need to make an argument as to why it's needed.
This function was introduced in WP 1.5.

**Comment by contactashish13** [comment]
@EkoJr I thinkhave_posts() will work on the main query when inside the Loop and not the $query object passed. Maybe we should be using $query->post_count?

**Comment by EkoJr** [comment]
@contactashish13 I was thinking the same as well with the possibility of there being an issue with the builtin function. 

Looking at it closer, I noticed there could be other implications that could alter the query. [WP_Query::have_posts()](https://developer.wordpress.org/reference/classes/wp_query/have_posts/). Plus, the function is usually part of The Loop in WP, and $query->post_count is still null in this case.

So, I would have to lean towards with **Method 1** on this.

**Comment by wpsmort** [comment]
The PR for this is #1522 and is ready for code review.

**Comment by michaeltorbert** [comment]
I'm removing the "completed testing," as the code has been redone, and testing will need to be redone once the development is finished.

**Comment by wpsmort** [comment]
@EkoJr Whats the status of this?  Has the code been redone as per the last comment?

**Comment by EkoJr** [comment]
PR #1595 

It has a travis fail due to a WP_UnitTest function missing during 4.0, but later found an annotation to require the function. So, I'll be adding 1 more commit to fix that for unit testing.

Other than that the plugin code would be ready for testing.

**Comment by bansay** [comment]
I'm still seeing this error. 

Warning: count(): Parameter must be an array or an object that implements Countable in /wp-content/plugins/all-in-one-seo-pack/aioseop_class.php

Running PHP 7.2
WP 4.6.4
All in One SEO version 2.3.9.1

**Comment by michaeltorbert** [comment]
@bansay Just so I'm clear, you're reporting that your AIOSEOP 2.3.9.1 still has a bug that, as you can see on this issue, we fixed in 2.5 over a year and a half ago?

**Comment by bansay** [comment]
My mistake. I didn't realize there was an update to the plugin. Somehow missed that when I looked. Sorry about that - this is fixed.
